MQTT logger V2

Blake-Haydon opened the pull request:
## Description
Adds robust MQTT logging and playback capabilities. This allows for all MQTT messages to be simulated at real time. Another step towards a fully virtual bike!

**This PR:**
✅ Logger is now general purpose 
✅ Playback _finally_ added  
✅ Proper exception handling

**Next PR:**
❌ Old logger needs to be removed
❌ Update ReadMe and fix file names `mqtt_playback_V2.py` -> `mqtt_playback.py`
❌ Add JSON un-wrapper to produce nice csv logs that play nice with excel



## Screenshots

`mqtt_playback_V2.py` verbose output -> (real time | topic | message) 
<img width="543" alt="image" src="https://user-images.githubusercontent.com/23159604/90332661-12e86d00-e002-11ea-8776-a491cfd8a0a1.png">


`mqtt_playback_V2.py` verbose output -> (real time | sped up time | topic | message) 
<img width="444" alt="image" src="https://user-images.githubusercontent.com/23159604/90332558-28a96280-e001-11ea-8c11-67c785e68224.png">



## Steps to Test

### Test the logger `mqtt_recorder_V2.py`
1. `cd  DAS`
2. `poetry shell`
3. Subscribe to sensor/# and test/# @ broker.hivemq.com for 10s
`python -m das.mqtt_recorder_V2 sensor/# test/#  -v --host broker.hivemq.com -v -t 10`
4. You should now see a log file in csv_data

### Test the playback `mqtt_playback_V2.py`
1. `cd  DAS`
2. `poetry shell`
3.  Playback 0001_log.csv at 30x speed using localhost MQTT broker
`python -m das.mqtt_playback_V2 das/csv_data/0001_log.csv -v -s 30`
4. You can check it works by running `mosquitto_sub -t "#"`

Blake-Haydon committed:
Add initial logger_V2

Blake-Haydon committed:
Add basic logging class

Blake-Haydon committed:
Add nice recording

Blake-Haydon committed:
Make sure there are no negative logs

Blake-Haydon committed:
Add correct ingestion

Blake-Haydon committed:
Add full playback functionality

Blake-Haydon committed:
Fix delimiter and comma to ensure that it works with all datatypes

Blake-Haydon committed:
Add proper command line input for playback

Blake-Haydon committed:
Add little -s for speed flag

Blake-Haydon committed:
Add scaled speed as an output for the logging

Blake-Haydon committed:
Rearange files

Blake-Haydon committed:
Clean redundant code

Blake-Haydon committed:
Fix logger command line tool

Blake-Haydon committed:
Add tests and use async for playback

Blake-Haydon committed:
Add nice async code

Blake-Haydon committed:
Change logger to Record to aid in readability

Blake-Haydon committed:
Add timed input for the logger

Blake-Haydon committed:
Add correct graceful exit for ctrl c

Blake-Haydon committed:
Rename logger to recorder

Blake-Haydon committed:
Add documentation for Record class

Blake-Haydon committed:
Add dict to store csv config

Blake-Haydon committed:
Cleanup old syncronous coude

Blake-Haydon committed:
Add inline comments for reading log file

Blake-Haydon committed:
Add docs for Playback class

Blake-Haydon committed:
Update poetry

hallgchris commented on DAS/das/mqtt_recorder_V2.py:
```diff
@@ -0,0 +1,56 @@
+                pass
+            except KeyboardInterrupt:
+                # Graceful exit
+                print(f"\nData saved in {CSV_FILEPATH}")
```
Can we print the name of the logfile here as well?

Blake-Haydon replied:
Yes it does this correctly now!

hallgchris commented on DAS/das/mqtt_recorder_V2.py:
```diff
@@ -0,0 +1,56 @@
+                pass
+            except KeyboardInterrupt:
+                # Graceful exit
+                print(f"\nData saved in {CSV_FILEPATH}")
```
I think it would make sense to print this inside `main_recorder.stop()`, so that
- We can make sure it only prints if writing the file was successful
- It prints also when we're running the logger for a fixed duration of time

Blake-Haydon replied:
Moved into the .stop() method

hallgchris commented on DAS/das/mqtt_recorder_V2.py:
```diff
@@ -0,0 +1,56 @@
+        while True:
+            try:
+                pass
+            except KeyboardInterrupt:
```
Rather than only doing this when manually closed with a `KeyboardInterrupt`, we probably want to at least *try* to write out the data in the case of any exception. Should we use a try-finally thing here?

hallgchris replied:
Any thoughts on this?

khanguslee replied:
> Should we use a try-finally thing here?

Sounds like a good idea to me!

Blake-Haydon replied:
I added a graceful_exit function which if an error occurs it will gracefully exit saving the last data

hallgchris commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+}
+
+
+class Record:
```
I feel like `Recorder` (or even `Logger`?) would be a better class name, the verb `Record` feels like more of a function name.

Blake-Haydon replied:
When you call the function it will be `logger.Record()`. It used to be `logger.Logger()` but I was getting confused when working on it. I'd be fine to change it to `Recorder` if others think its better

hallgchris replied:
`Record` isn't really a function though, it's a class. This becomes more obvious if you import it differently, so that it's not prefixed with the module name:
```python
from das.utils.logger import Record()
main_recorder = Record()
```
Considering the variable you're assigning the class to also contains `recorder`, I do kinda feel like the class should be called that and the assignment would make more sense. Happy to hear more opinions though!

khanguslee replied:
Had a good think about this. After reading the implementation inside `mqtt_recorder_v2.py` I think it would be good to use it like:

```python
    main_recorder = logger.Recorder(
        CSV_FILEPATH, topics=args.topics, broker_address=args.host, verbose=args.verbose
    )
	main_recoder.start() # or main_recoder.play()
```

This is to match with the `main_recorder.stop()`. It's not quite intuitive to know when we start recording. This should also solve the problem of `Record` vs `Recorder` too.

Blake-Haydon replied:
Totally agree, added a start function. It makes it significantly easier to work with!

hallgchris commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+
+    Attributes
+    ----------
+    TOPICS: List(str)
```
What differentiates an attribute in capitals vs lowercase?

Blake-Haydon replied:
Capitals were meant to be const. Do you reckon it just makes it more confusing?

hallgchris replied:
Yeah that makes sense! Just hadn't really seen it in member variables before but I see nothing wrong with it. Should `_client` also be in capitals then?

hallgchris commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+        for filename in os.listdir(csv_folder_path):
+            try:
+                if int(filename[0:4]) > previous_log_num:
+                    previous_log_num = abs(int(filename[0:4]))
```
Why absolute value?

Blake-Haydon replied:
I was thinking it could potentially be a negative filename which would mess it up with the hardcoding. It has been changed to a regular expression.

hallgchris commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+            except ValueError:
+                if self._VERBOSE:
+                    print(
+                        f"WARNING: {filename} should not be in {csv_folder_path}")
```
Maybe make a `log` function which checks `self._VERBOSE`, it's quite a common pattern in this file. You could make the function shared by both `Record` and `Playback` by making some superclass they both extend which contains the function.

hallgchris commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+        )
+
+        # Add headers for csv
+        self._LOG_FILE_WRITER.writeheader()
```
I reckon once we've been able to successfully write to the file (here), it would be good to say something like "Writing to logfile <filepath>", just because when running in non-verbose, there's very little feedback as far as whether things are working.

hallgchris commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+        """
+
+        if self._VERBOSE:
+            print(f"⚡ Playback initiated at {speed}x speed ⚡")
```
I think we could print this even when not on verbose mode. Otherwise we don't see anything, and I think some form of feedback that it's working would be good!

hallgchris commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+        self._log_data = []
+        for row in csv_reader:
+            # Convert time delta to float
+            row["time_delta"] = float(row["time_delta"])
```
Could you get these column headers from `CsvConfig`? If that's too hard to read, maybe you could define constants for the header titles, and use the constants both here and in `CsvConfig`.

hallgchris commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+            publish_queue.append(_publish_aux(row))
+
+        # Load all async operation at the start using gather
+        await asyncio.gather(*publish_queue, return_exceptions=True)
```
This async stuff is really cool, but is there any reason we need it? I don't see how the way it's currently coded has any benefit over looking at each row iteratively, unless you feel that for some reason we need to support logs where the rows don't appear in chronological order.

I do appreciate the coolness though!

Blake-Haydon replied:
I originally implemented it without async using `time.sleep`. I expected there would be a small difference between playback and actual time. Turns out there was a difference of 1s every minute. So for an hour recording we would be about 1m behind if we played it back.

The implementation in async is also much easier to understand and requires less math.

hallgchris replied:
It could be implemented fairly easily with `time.sleep`, just by storing the time which the last message was sent and then calculating how long until the time the next message should be sent. I think my main concern with async is that spawning thousands of threads feels messy and could effect performance (maybe, unsure) and feels unnecessary when the same logic could be achieved with a loop that people newer to coding would understand.

hallgchris commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+        MQTT client that connects to the broker and recives the messages
+    """
+
+    def __init__(self, csv_folder_path: str, topics: list = ["#"], broker_address: str = "localhost", verbose: bool = False) -> None:
```
Some of these lines are super long - even if we don't go for a hard 79 character limit, could you split these huge ones up a bit?

hallgchris commented on DAS/das/mqtt_playback_V2.py:
```diff
@@ -0,0 +1,34 @@
+
+parser.add_argument(
+    '--host', action='store', type=str, default="localhost",
+    help="""Address of the MQTT broker. If nothing is selected it will
```
Nit, but "If no broker is specified" might make slightly more sense.

khanguslee replied:
Hmm, that's weird. I always thought `argparse` would show you the default values when you type `--help`. Turns out you need to add the default arg formatter. seen here: https://stackoverflow.com/a/12151325/10947824

Can you implement that into your script and remove all strings that mention what the default is.

We'll need to do that in all scripts that we have in MHP later too..

hallgchris commented on DAS/pyproject.toml:
```diff
@@ -10,6 +10,8 @@ paho-mqtt = "^1.5.0"
 pandas = "^1.0.3"
 argparse = "^1.4.0"
+pathlib = "^1.0.1"
+asyncio = "^3.4.3"
```
These are part of standard python aren't they? Why do they need to be in `pyproject.toml`?

hallgchris commented on DAS/das/tests/test_logger.py:
```diff
@@ -0,0 +1,36 @@
+
+    main_logger = logger.Record(TEST_FOLDER)
+    main_playback = logger.Playback(
+        os.path.join(DATA_FOLDER, "test_log_1.csv"))
```
You may have forgotten to upload this file. Maybe rather than having a random file sitting around in the repo, you could make this function generate the file?

Blake-Haydon replied:
Really good idea!

hallgchris commented on DAS/das/tests/test_logger.py:
```diff
@@ -0,0 +1,36 @@
+from das.utils import logger
```
We've got to figure these folders out... Despite this being in a `tests` folder, this is the only thing that's a unit test, the rest I guess are more like testing utilities?

khanguslee replied:
Whatever isn't a test inside the `tests` folder should relocated into the utils folder for now I reckon. (different PR)

Blake-Haydon replied:
Yeah definitely need to move them into utils

hallgchris reviewed (changes requested):
This looks really great! Definitely agree it's neater than the previous logger.

I'm curious as to how you envision this being used on the bike. If it were always running in the background, everything would be in one log file. Is this what you intend? Cos I suppose the file could be chopped up in the JSON parsing stage you commented about. Otherwise we'd need to have something to start and terminate the `Record` class, which actually is entirely doable now I think about it.

I like your idea with the JSON parsing producing a new CSV after the fact. Makes the asynchronous nature of separate sensor modules easier to deal with. When this is done, we should make this occur automatically on the DAS so that we can still download useful logs from the das immediately after the ride so we can get right into seeing how fast we went!

khanguslee commented on DAS/das/tests/test_logger.py:
```diff
@@ -0,0 +1,36 @@
+TEST_FOLDER = os.path.join(CURRENT_FILEPATH, "csv_data")
+
+
+def test_Logger_make_csv_folder():
```
Great that you wrote unit tests! I don't think we've set up `pytest` in this repo yet. Can your next PR be setting up `pytest` so it runs your unit tests and also a Github Action to run them with every PR (see other repos for examples).

khanguslee commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+        MQTT client that connects to the broker and recives the messages
+    """
+
+    def __init__(self, csv_folder_path: str, topics: list = ["#"], broker_address: str = "localhost", verbose: bool = False) -> None:
```
Think this line definitely is above the 79 characters per line limit LOL

Blake-Haydon replied:
haha, just a few over. Is there any good code formatters for python. Is black any good?

hallgchris replied:
Yeah black seems to work well!

khanguslee commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+    """
+
+    def __init__(self, csv_folder_path: str, topics: list = ["#"], broker_address: str = "localhost", verbose: bool = False) -> None:
+        # The logger object can subscribe to many topics (if none are selected then it will subscrive to all)
```
subscrive -> subscribe

khanguslee commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+
+        # Name the csv log file xxxx_log.csv where xxxx is a number
+        previous_log_num = 0
+        for filename in os.listdir(csv_folder_path):
```
Might be better to abstract this out into a private method so the constructor looks cleaner.

khanguslee commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+            for topic in self.TOPICS:
+                self._client.subscribe(topic)
+                if self._VERBOSE:
+                    print(f"Subscribed to: {topic}")
```
Another possible future refactor for these scripts is to implement an actual logger so we don't have any `print()` statements and can properly display different levels of information.

Blake-Haydon replied:
Just added the real python logger.

I looked into using it to do the real logging functionality, but its not designed to record data. It is only good at logging human readable output.

khanguslee replied:
Yeah i'm happy for it to be used as a way to log our debug statements for now!

khanguslee commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,235 @@
+        asyncio.run(self._publish(speed))
+
+    async def _publish(self, speed):
+        """Async function that collects all the necessary publish functions and gathers them to then be run by the
```
Be careful with the long lines.

khanguslee reviewed (changes requested):
Yup just to clarify, we now have all topics recorded into a single file. I am fine with this, it just means we need to create a new python script that splits our CSV file into individual wireless modules via the topics. I actually like this format a lot better than previously as it is much easier to replay logs.

We might need to format this repo soon so it conforms with the other repos.. Anyone want to volunteer?

Blake-Haydon committed:
Add proper logger

Blake-Haydon committed:
Add graceful exit function

Blake-Haydon committed:
Fix order of logging for exit

Blake-Haydon committed:
Allow for unlimited logs

Blake-Haydon committed:
Remove old comment

Blake-Haydon committed:
Move function out of recorder init

Blake-Haydon committed:
Make non vervose more verbose

Blake-Haydon committed:
Fix parser.add_argument for broker

Blake-Haydon committed:
Update poetry

Blake-Haydon committed:
Add default for argparse

Blake-Haydon committed:
Fix spelling mistake subscribe

Blake-Haydon committed:
Add self contained tests

Blake-Haydon committed:
Black formatting

hallgchris commented on DAS/das/tests/test_logger.py:
```diff
@@ -1,36 +1,149 @@
 import shutil
 import os
+import time
+import unittest
```
Any reason for going with `unittest` over `pytest`? Not totally against it! But asking because it might be better to be consistent with the rest of our repos which use `pytest`.

hallgchris replied:
Also how do you run the tests? :sweat_smile: 

Would be good to document in readme

Blake-Haydon replied:
You run tests by just writing `pytest`. `pytest` can be used as just a runner for ordinary `unittest` tests. I used `unittest` as it requires building up a test and then tearing it down and its much easier to use `unittest`.

hallgchris replied:
Oh right, did not know you could do that!

hallgchris reviewed (changes requested):
This is looking really good! Great unit tests and I like the `logging` library, maybe we need to use it elsewhere too. Not much more to add other than the stuff above!

hallgchris reviewed (dismissed):
Sorry this took a while! Assuming you've made sure you've looked through my prior replies and don't agree with them (which is fine, maybe drop a reply as to why) then happy to approve :)

khanguslee commented on DAS/das/mqtt_playback_V2.py:
```diff
@@ -0,0 +1,47 @@
+    action="store",
+    type=float,
+    default=1,
+    help="""Playback speed up. Default is 1x""",
```
Can remove `Default is 1x` since you added the default formatter in argparse:

![Screen Shot 2020-09-05 at 4 01 37 pm](https://user-images.githubusercontent.com/18709969/92298915-15027380-ef91-11ea-8e04-3d9b96cd28f4.png)

khanguslee reviewed (changes requested):
SO CLOSEEEE. Added a few comments to existing discussions.

Blake-Haydon committed:
Change Record class to Recorder class

Blake-Haydon committed:
uppercase _CLIENT property

Blake-Haydon committed:
Add type annotations to return values

Blake-Haydon committed:
Add start recording method

Blake-Haydon committed:
Remove double default text

Blake-Haydon committed:
Update test files to new Recorder class

Blake-Haydon commented:
I think everything is looking good now!

Blake-Haydon committed:
Add general exception handling case

Blake-Haydon committed:
Refactor graceful exit

Blake-Haydon committed:
Add gracefull exit and to playback tool

khanguslee commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,275 @@
+        self._CLIENT.loop_start()  # Threaded execution loop
+
+    def _create_log_file(self, csv_folder_path: str) -> None:
+        """Is used to open a log file and csv obj during the object init
```
nit: Imperative mood for any method description btw!

Example:
`Generates a log file ready to be written in.`

khanguslee commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,275 @@
+    ) -> None:
+        # The logger object can subscribe to many topics (if none are selected then it will subscribe to all)
+        self.TOPICS = topics
+        self._VERBOSE = verbose
```
Instead of having the `if self._VERBOSE:` everywhere, you can actually tell the `logger` library to hide certain levels of logging (so can automatically hide all the info and debug logs).

``` python
if (verbose):
    logging.getLogger().setLevel(logging.ERROR)
```

I think the above might work. pls test.

Blake-Haydon replied:
Works a treat!

khanguslee commented on DAS/das/utils/logger.py:
```diff
@@ -0,0 +1,275 @@
+                )
+
+            try:
+                publish.single(
```
Kinda inclined to change the alias for `publish` to `mqtt_publish` or `publish_mqtt` (or something similar). I wasn't sure where this function came from since it's pretty similar to `_publish`.

Blake-Haydon replied:
I've switched to using a client instead of just a single publish statement. This resolves this issue while also making the error messages much more clear!

khanguslee reviewed (changes requested):
I'm pretty happy overall. I've tested it and it works amazingly too!

Just a few things that you probably didn't know about logging, but otherwise it's on the right track!

hallgchris commented on DAS/das/mqtt_recorder_V2.py:
```diff
@@ -52,22 +52,25 @@
-            except KeyboardInterrupt:
-                graceful_exit(main_recorder)
+
+    except (KeyboardInterrupt, Exception):
+        pass
```
This means if an exception is thrown we won't ever get to see what happened. Maybe something like
```python
    except KeyboardInterrupt:
        pass
    except Exception as e:
        print(f"{type(e)}: {e}")
```
I know we're not likely to have anything go wrong in that try section but I'm just a little scared by catching all exceptions and ignoring them.

Blake-Haydon replied:
100% agree will implement!

hallgchris reviewed (changes requested):
Just one more thing to add on to what Angus said! We're nearly there

Blake-Haydon committed:
Add better error handling and remove self._LOGGER

Blake-Haydon commented:
Only 3 main updates from the previous version.
- Imperative mood for any method description (checked and all fixed)
- Removed `self._VERBOSE` everywhere (much cleaner)
- Added more exceptions to catch when it doesn't connect to MQTT broker

Blake-Haydon commented:
Oh and all the tests work when running pytest 🎉

hallgchris reviewed (approved):
LGTM! Thanks for this!

khanguslee reviewed (approved):
ayyy LGTM! Looks good now 💯
